Kubernetes/legos/k8s_get_memory_utilization_of_services/k8s_get_memory_utilization_of_services.py

```diff
@@ -131,38 +124,72 @@ def k8s_get_memory_utilization_of_services(handle, namespace: str = "", threshol
-
-        for pod, mem_usage in pods_to_check.items():
-            # Lets get memory requests for all containers not just the first container
-            kubectl_command = f"kubectl get pod {pod} -n {namespace} -o=jsonpath='{{.spec.containers[*].resources.requests.memory}}'"
-            response = handle.run_native_cmd(kubectl_command)
-            mem_request = response.stdout.strip()
-            _mem_request = 0
-            for _mr in mem_request.split():
-                # Lets calculate the total memory request for all containers in the POD
-                _mem_request += convert_memory_to_bytes(_mr)
-            mem_request = _mem_request
-            mem_usage = convert_memory_to_bytes(mem_usage)
-
-            if not mem_request:
-                # Memory limit is not set, dont calculate utilization
-                continue
-            utilization = (mem_usage / mem_request) * 100
-            utilization = round(utilization, 2)
-            if utilization > threshold:
-                exceeding_services.append({
-                    "pod": pod,
-                    "namespace": namespace,
-                    "utilization_percentage": utilization,
-                    "memory_request_bytes": mem_request,
-                    "memory_usage_bytes": mem_usage,
-                })
+                    top_pod_containers_cmd = f"kubectl top pod {svc_pod} -n {namespace} --containers --no-headers"
+                    response = handle.run_native_cmd(top_pod_containers_cmd)
+                    top_pod_containers_output = response.stdout.strip()
+                    if not top_pod_containers_output:
+                        continue
+
+                    for line in top_pod_containers_output.split('\n'):
+                        parts = line.split()
+                        if len(parts) >= 3:  # Ensure line has enough parts to parse
+                            container_name = parts[1]
+                            mem_usage = parts[-1]
+
+                        # Key: Service, Pod, Container; Value: Memory Usage
+                        service_pods_containers[(svc, svc_pod, container_name)] = mem_usage
+        else:
+            # Fetch metrics for all pods in the namespace if no specific services are provided
+            top_pods_command = f"kubectl top pods -n {namespace} --containers --no-headers"
+            response = handle.run_native_cmd(top_pods_command)
+            top_pods_output = response.stdout.strip()
+            if not top_pod_containers_output:
+                return (True, None)
+
+            for line in top_pods_output.split('\n'):
+                parts = line.split()
+                if len(parts) >= 3:
+                    pod_name, container_name, mem_usage = parts[0], parts[1], parts[-1]
+
+                # Key: Service: None, Pod, Container; Value: Memory Usage (when services are not specified)
+                service_pods_containers[(None, pod_name, container_name)] = mem_usage
```

Comment:
Why did we set service to `None`?

In the printer we will print N/A for all pods, which makes it look like the `pod` it not part of a service (which is a valid configuration).

I'm not sure about customer configuration, but we could make `service` parameter mandatory.

Reply:
In case when only namespace is given, we are fetching all pods in that namespace and hence setting the Service to None. 

In the final exceeding_services list, when the services are specified ( which is for lb) all service names will be shown but when no services are given and we have only namespace, we are not looking up which service it belongs to and hence is set to None. 

IMO having services parameter mandatory would be ideal.

Reply:
Have made service (and namespace) as mandatory parameter. Because if service is given we also need namespace where the service is configured.